torch/csrc/jit/codegen/cuda/predicate_compute.cpp

```diff
@@ -306,35 +229,28 @@ void UnswitchPredicate::predicateOn(kir::Expr* tv_expr) {
 
   auto out_tv = firstTensorViewOutput(tv_expr);
 
-  // For the case of generating predicates, it's safe to assume all
-  // axes are contiguous and saves some redundant predicates.
-  auto pred_contiguity =
-      std::vector<bool>(out_tv->domain()->rootDomain().size(), true);
+  auto pred_info = Index::getReferenceRootPredicates(out_tv, for_loops_, true);
 
-  auto pred_inds = Index::getConsumerRootPredIndices(
-      out_tv, for_loops_, pred_contiguity, true);
-  auto root_indices = pred_inds.first;
-  auto use_rfactor = pred_inds.second;
+  for (auto i : c10::irange(pred_info.first.size())) {
+    auto pred = pred_info.first[i];
+    const auto& root_ids = pred_info.second[i];
 
-  auto all_preds =
-      PredicateCompute::computePredicates(out_tv, root_indices, use_rfactor);
+    bool add_pred = false;
 
-  const auto out_domain = out_tv->domain();
-  const auto root_dom = (use_rfactor && out_domain->hasRFactor())
-      ? out_domain->rfactorDomain()
-      : out_domain->rootDomain();
+    for (auto root_id : root_ids) {
+      auto kir_root_id = gpu_lower->lowerValue(root_id)->as<kir::IterDomain>();
 
-  TORCH_INTERNAL_ASSERT(
-      all_preds.size() == root_dom.size(),
-      "Predicates should be produced for every dimension, even if it's simply set as true.");
-
-  for (size_t i = 0; i < all_preds.size(); i++) {
-    if (all_preds[i]->isConst() && all_preds[i]->value().value()) {
-      continue;
+      if (std::find(
+              predicated_iter_dom_.begin(),
+              predicated_iter_dom_.end(),
+              kir_root_id) == predicated_iter_dom_.end()) {
+        add_pred = true;
+        predicated_iter_dom_.push_back(kir_root_id);
+      }
+      if (add_pred) {
+        predicates_.push_back(pred);
```

Comment:
Don't you need to break here? Otherwise, it seems that `pred` could be pushed back to `predicates_` multiple times.